src/NLog.Wcf/LogEventInfoBuffer.cs

```diff
@@ -0,0 +1,155 @@
+        /// <param name="growLimit">The maximum number of items that the buffer can grow to.</param>
+        public LogEventInfoBuffer(int size, bool growAsNeeded, int growLimit)
+        {
+            _growAsNeeded = growAsNeeded;
+            _buffer = new AsyncLogEventInfo[size];
+            _growLimit = growLimit;
+            _getPointer = 0;
+            _putPointer = 0;
+        }
+
+        /// <summary>
+        /// Gets the capacity of the buffer
+        /// </summary>
+        public int Size => _buffer.Length;
+
+        /// <summary>
+        /// Gets the number of items in the buffer
+        /// </summary>
+        internal int Count { get { lock (_lockObject) return _count; } }
+
+        /// <summary>
+        /// Adds the specified log event to the buffer.
+        /// </summary>
+        /// <param name="eventInfo">Log event.</param>
+        /// <returns>The number of items in the buffer.</returns>
+        public int Append(AsyncLogEventInfo eventInfo)
+        {
+            lock (_lockObject)
+            {
+                // make room for additional item
+                if (_count >= _buffer.Length)
+                {
+                    if (_growAsNeeded && _buffer.Length < _growLimit)
+                    {
+                        // create a new buffer, copy data from current
+                        int newLength = _buffer.Length * 2;
+                        if (newLength >= _growLimit)
+                        {
+                            newLength = _growLimit;
+                        }
+
+                        InternalLogger.Trace("Enlarging LogEventInfoBuffer from {0} to {1}", _buffer.Length, newLength);
+                        var newBuffer = new AsyncLogEventInfo[newLength];
+                        Array.Copy(_buffer, 0, newBuffer, 0, _buffer.Length);
+                        _buffer = newBuffer;
+                    }
+                    else
+                    {
+                        // lose the oldest item
+                        _getPointer = _getPointer + 1;
+                    }
+                }
+
+                // put the item
+                _putPointer = _putPointer % _buffer.Length;
+                _buffer[_putPointer] = eventInfo;
+                _putPointer = _putPointer + 1;
+                _count++;
+                if (_count >= _buffer.Length)
+                {
+                    _count = _buffer.Length;
+                }
+
+                return _count;
+            }
+        }
+
+        /// <summary>
+        /// Gets the array of events accumulated in the buffer and clears the buffer as one atomic operation.
+        /// </summary>
+        /// <returns>Events in the buffer.</returns>
+        public AsyncLogEventInfo[] GetEventsAndClear()
+        {
+            lock (_lockObject)
+            {
+                int cnt = _count;
+                if (cnt == 0)
+                    return new AsyncLogEventInfo[0];
```

Comment:
Had to change here as well. use of `ArrayHelper` which is internal in NLog/Common
@snakefoot

Reply:
I have no problem with removing use of ArrayHelper. This will work just fine.